tests/test_casefold_migration.py

```diff
@@ -0,0 +1,335 @@
+import json
+import os.path
+from unittest.mock import patch
+
+from twisted.trial import unittest
+
+from scripts.casefold_db import (
+    calculate_lookup_hash,
+    update_global_assoc,
+    update_local_associations,
+)
+from sydent.util import json_decoder
+from sydent.util.emailutils import sendEmail
+from tests.utils import make_sydent
+
+
+class MigrationTestCase(unittest.TestCase):
+    def create_signedassoc(self, medium, address, mxid, ts, not_before, not_after):
+        return {
+            "medium": medium,
+            "address": address,
+            "mxid": mxid,
+            "ts": ts,
+            "not_before": not_before,
+            "not_after": not_after,
+        }
+
+    def setUp(self):
+        # Create a new sydent
+        config = {
+            "general": {
+                "templates.path": os.path.join(
+                    os.path.dirname(os.path.dirname(__file__)), "res"
+                ),
+            },
+            "crypto": {
+                "ed25519.signingkey": "ed25519 0 FJi1Rnpj3/otydngacrwddFvwz/dTDsBv62uZDN2fZM"
+            },
+        }
+        self.sydent = make_sydent(test_config=config)
+
+        # create some local associations
+        associations = []
+
+        for i in range(15):
```

Comment:
Maybe specify here which address range we expect to conflict? Otherwise it feels a bit confusing.
Or better, do it with 2 loops (one with `range(10)` and the other with `range(5)`). In test, we want to optimise for readability, so it's fine if the code doesn't look as performant as you would have it in the rest of the codebase.